interfaces/tests/sync_int/sync_int.c

```diff
@@ -0,0 +1,211 @@
+/*Codes_SRS_SYNC_43_008: [wait_on_address shall wait indefinitely until it is woken up by a call to wake_by_address_[single/all] if timeout_ms is equal to UINT32_MAX]*/
+/*Codes_SRS_SYNC_43_003: [ wait_on_address shall wait until another thread in the same process signals at address using wake_by_address_[single/all] and return true. ]*/
+/*Codes_SRS_SYNC_43_004: [ wake_by_address_all shall cause all the thread(s) waiting on a call to wait_on_address with argument address to continue execution. ]*/
+/*Codes_SRS_SYNC_43_005: [ wake_by_address_single shall cause one thread waiting on a call to wait_on_address with argument address to continue execution. ]*/
+TEST_FUNCTION(wake_up_all_threads)
+{
+    ///arrange
+    volatile_atomic int32_t var;
+    interlocked_exchange(&var, 0);
+    interlocked_exchange(&create_count, 0);
+    THREAD_HANDLE threads[100];
+
+    ///act
+    for(int i = 0; i < 100; ++i){
+        ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK, ThreadAPI_Create(&threads[i], increment_on_wake_up, (void*)&var));
+    }
+
+    int current_create_count = interlocked_add(&create_count, 0);
+    while (current_create_count < 100)
+    {
+        wait_on_address(&create_count, &current_create_count, UINT32_MAX);
+        current_create_count = interlocked_add(&create_count, 0);
+    }
+    wake_by_address_all(&var);
+    for(int i = 0; i < 100; ++i){
+        int return_val;
+        ThreadAPI_Join(threads[i], &return_val);
+    }
+    
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, 100, interlocked_add(&var, 0), "Return value is incorrect");
+}
+
+
+/*Codes_SRS_SYNC_43_001: [ wait_on_address shall atomically compare *address and *compare_address.]*/
+/*Codes_SRS_SYNC_43_002: [ wait_on_address shall immediately return true if *address is not equal to *compare_address.]*/
+TEST_FUNCTION(wait_on_address_returns_immediately)
+{
+    ///arrange
+    volatile_atomic int32_t var;
+    interlocked_exchange(&var, 0);
+    int value = 1;
+
+    ///act
+    bool return_val = wait_on_address(&var, &value, UINT32_MAX);
+
+    ///assert
+    ASSERT_IS_TRUE(return_val, "wait_on_address should have returned true");
+}
+
+
+/*Codes_SRS_SYNC_43_001: [ wait_on_address shall atomically compare *address and *compare_address.]*/
+/*Codes_SRS_SYNC_43_002: [ wait_on_address shall immediately return true if *address is not equal to *compare_address.]*/
+/*Codes_SRS_SYNC_43_009: [ If timeout_ms milliseconds elapse, wait_on_address shall return false. ]*/
+TEST_FUNCTION(wait_on_address_returns_after_timeout_elapses)
+{
+    ///arrange
+    volatile_atomic int32_t var;
+    interlocked_exchange(&var, 0);
+    int value = 0;
+    int timeout = 1000;
+    /// act
+    double start_time = timer_global_get_elapsed_ms();
+    bool return_val = wait_on_address(&var, &value, timeout);
+    double time_elapsed = timer_global_get_elapsed_ms() - start_time;
+
+    ///assert
+    ASSERT_IS_TRUE(time_elapsed < timeout * 1.5, "Too much time elapsed");
```

Comment:
>elapsed [](start = 64, length = 7)

in this case, would be good to print the value of `time_elapsed` (this works with printf syntax btw).